File: wfm-develop mobile 07.05.2026/wfm/mobile/ios/WFMApp/Core/Analytics/FirebaseAnalyticsService.swift
import Foundation
import FirebaseAnalytics

/// Firebase реализация аналитики.
///
/// Подключение:
/// 1. Добавить пакет Firebase через Xcode: File → Add Package Dependencies →
///    https://github.com/firebase/firebase-ios-sdk
///    Выбрать только FirebaseAnalytics.
/// 2. Добавить GoogleService-Info.plist в таргет WFMApp.
/// 3. В DependencyContainer заменить NoOpAnalyticsService на FirebaseAnalyticsService.
/// 4. В WFMApp.swift добавить FirebaseApp.configure() до создания DependencyContainer.
final class FirebaseAnalyticsService: AnalyticsService {

    func track(_ event: AnalyticsEvent) {
        let (name, params) = event.firebaseRepresentation
        Analytics.logEvent(name, parameters: params.isEmpty ? nil : params)
    }

    func setUser(userId: Int, role: String) {
        Analytics.setUserID(String(userId))
        Analytics.setUserProperty(role, forName: "user_role")
    }

    func resetUser() {
        Analytics.setUserID(nil)
        Analytics.setUserProperty(nil, forName: "user_role")
    }
}

// MARK: - Firebase mapping

private extension AnalyticsEvent {
    var firebaseRepresentation: (name: String, params: [String: Any]) {
        switch self {
        case .phoneInputViewed:
            return ("phone_input_viewed", [:])
        case .phoneSubmitted:
            return ("phone_submitted", [:])
        case .codeInputViewed:
            return ("code_input_viewed", [:])
        case .codeSubmitted:
            return ("code_submitted", [:])
        case .loginCompleted(let role):
            return ("login_completed", ["role": role])
        case .logoutTapped:
            return ("logout_tapped", [:])
        case .homeViewed:
            return ("home_viewed", [:])
        case .tasksListViewed(let count):
            return ("tasks_list_viewed", ["tasks_count": count])
        case .taskCardTapped(let state, let type):
            return ("task_card_tapped", ["task_state": state, "task_type": type])
        case .taskDetailViewed(let state, let review, let type):
            return ("task_detail_viewed", ["task_state": state, "task_review_state": review, "task_type": type])
        case .taskStartTapped(let taskId, let taskType, let workType, let timeStart, let timeEnd):
            var params: [String: Any] = ["task_id": taskId, "task_type": taskType]
            if let workType = workType { params["work_type"] = workType }
            if let timeStart = timeStart { params["time_start"] = timeStart }
            if let timeEnd = timeEnd { params["time_end"] = timeEnd }
            return ("task_start_tapped", params)
        case .taskPauseTapped:
            return ("task_pause_tapped", [:])
        case .taskResumeTapped:
            return ("task_resume_tapped", [:])
        case .taskCompleteSheetOpened(let requires):
            return ("task_complete_sheet_opened", ["requires_photo": requires])
        case .taskCompleteSubmitted(let hasPhoto, let hasText, let taskId, let taskType, let workType, let timeStart, let timeEnd):
            var params: [String: Any] = ["has_photo": hasPhoto, "has_text": hasText, "task_id": taskId, "task_type": taskType]
            if let workType = workType { params["work_type"] = workType }
            if let timeStart = timeStart { params["time_start"] = timeStart }
            if let timeEnd = timeEnd { params["time_end"] = timeEnd }
            return ("task_complete_submitted", params)
        case .managerHomeViewed(let count):
            return ("manager_home_viewed", ["tasks_on_review_count": count])
        case .taskReviewSheetOpened(let reviewState):
            return ("task_review_sheet_opened", ["task_review_state": reviewState])
        case .taskApprovedTapped:
            return ("task_approved_tapped", [:])
        case .taskRejectedTapped(let hasComment):
            return ("task_rejected_tapped", ["has_comment": hasComment])
        case .managerTasksViewed:
            return ("manager_tasks_viewed", [:])
        case .employeeFilterViewed:
            return ("employee_filter_viewed", [:])
        case .employeeFilterApplied(let count):
            return ("employee_filter_applied", ["selected_count": count])
        case .taskCreateSheetOpened:
            return ("task_create_sheet_opened", [:])
        case .taskCreateSubmitted(let hasAssignee, let requiresPhoto, let minutes):
            return ("task_create_submitted", ["has_assignee": hasAssignee, "requires_photo": requiresPhoto, "planned_minutes": minutes])
        case .taskEditSheetOpened:
            return ("task_edit_sheet_opened", [:])
        case .shiftOpenCompleted(let shiftId, let role):
            return ("shift_open_completed", ["shift_id": shiftId, "role": role])
        case .shiftCloseCompleted(let shiftId, let role):
            return ("shift_close_completed", ["shift_id": shiftId, "role": role])
        case .settingsViewed:
            return ("settings_viewed", [:])
        case .noAssignmentViewed:
            return ("no_assignment_viewed", [:])
        case .apiError(let status):
            return ("api_error", ["http_status": status])
        case .apiRequestCompleted(let path, let method, let httpStatus, let durationMs, let storeId, let isError, let errorType):
            var params: [String: Any] = [
                "path": path,
                "method": method,
                "http_status": httpStatus,
                "duration_ms": durationMs,
                "store_id": storeId,
                "is_error": isError
            ]
            if let errorType = errorType {
                params["error_type"] = errorType
            }
            return ("api_request_completed", params)
        case .pushNotificationReceived(let channel, let notificationId, let taskId):
            var params: [String: Any] = ["channel": channel]
            if let notificationId { params["notification_id"] = notificationId }
            if let taskId { params["task_id"] = taskId }
            return ("push_notification_received", params)
        case .pushNotificationTapped(let channel, let notificationId, let taskId):
            var params: [String: Any] = ["channel": channel]
            if let notificationId { params["notification_id"] = notificationId }
            if let taskId { params["task_id"] = taskId }
            return ("push_notification_tapped", params)
        case .subtaskSelectSheetOpened(let trigger, let workType, let operationsCount):
            var params: [String: Any] = ["trigger": trigger, "operations_count": operationsCount]
            if let workType { params["work_type"] = workType }
            return ("subtask_select_sheet_opened", params)
        case .subtaskSearchUsed:
            return ("subtask_search_used", [:])
        case .subtaskCreateSheetOpened:
            return ("subtask_create_sheet_opened", [:])
        case .subtaskCreated:
            return ("subtask_created", [:])
        case .hintsTabViewed(let workType, let zone, let hintsCount):
            var params: [String: Any] = ["hints_count": hintsCount]
            if let workType { params["work_type"] = workType }
            if let zone { params["zone"] = zone }
            return ("hints_tab_viewed", params)
        }
    }
}

File: wfm-develop mobile 07.05.2026/wfm/mobile/android/app/src/gms/kotlin/com/beyondviolet/wfm/core/analytics/FirebaseAnalyticsService.kt
package com.beyondviolet.wfm.core.analytics

import com.google.firebase.Firebase
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.analytics
import com.google.firebase.analytics.logEvent

/**
 * Реализация аналитики через Firebase Analytics.
 */
class FirebaseAnalyticsService : AnalyticsService {

    private val firebaseAnalytics: FirebaseAnalytics = Firebase.analytics

    override fun track(event: AnalyticsEvent) {
        firebaseAnalytics.logEvent(event.name()) {
            event.params().forEach { (key, value) -> param(key, value) }
        }
    }

    override fun setUser(userId: Int, role: String) {
        firebaseAnalytics.setUserId(userId.toString())
        firebaseAnalytics.setUserProperty("user_role", role)
    }

    override fun resetUser() {
        firebaseAnalytics.setUserId(null)
        firebaseAnalytics.setUserProperty("user_role", null)
    }

    private fun AnalyticsEvent.name(): String = when (this) {
        is AnalyticsEvent.PhoneInputViewed -> "phone_input_viewed"
        is AnalyticsEvent.PhoneSubmitted -> "phone_submitted"
        is AnalyticsEvent.CodeInputViewed -> "code_input_viewed"
        is AnalyticsEvent.CodeSubmitted -> "code_submitted"
        is AnalyticsEvent.LoginCompleted -> "login_completed"
        is AnalyticsEvent.LogoutTapped -> "logout_tapped"
        is AnalyticsEvent.HomeViewed -> "home_viewed"
        is AnalyticsEvent.TasksListViewed -> "tasks_list_viewed"
        is AnalyticsEvent.TaskCardTapped -> "task_card_tapped"
        is AnalyticsEvent.TaskDetailViewed -> "task_detail_viewed"
        is AnalyticsEvent.TaskStartTapped -> "task_start_tapped"
        is AnalyticsEvent.TaskPauseTapped -> "task_pause_tapped"
        is AnalyticsEvent.TaskResumeTapped -> "task_resume_tapped"
        is AnalyticsEvent.TaskCompleteSheetOpened -> "task_complete_sheet_opened"
        is AnalyticsEvent.TaskCompleteSubmitted -> "task_complete_submitted"
        is AnalyticsEvent.ManagerHomeViewed -> "manager_home_viewed"
        is AnalyticsEvent.TaskReviewSheetOpened -> "task_review_sheet_opened"
        is AnalyticsEvent.TaskApprovedTapped -> "task_approved_tapped"
        is AnalyticsEvent.TaskRejectedTapped -> "task_rejected_tapped"
        is AnalyticsEvent.ManagerTasksViewed -> "manager_tasks_viewed"
        is AnalyticsEvent.EmployeeFilterViewed -> "employee_filter_viewed"
        is AnalyticsEvent.EmployeeFilterApplied -> "employee_filter_applied"
        is AnalyticsEvent.TaskCreateSheetOpened -> "task_create_sheet_opened"
        is AnalyticsEvent.TaskCreateSubmitted -> "task_create_submitted"
        is AnalyticsEvent.TaskEditSheetOpened -> "task_edit_sheet_opened"
        is AnalyticsEvent.ShiftOpenCompleted -> "shift_open_completed"
        is AnalyticsEvent.ShiftCloseCompleted -> "shift_close_completed"
        is AnalyticsEvent.SettingsViewed -> "settings_viewed"
        is AnalyticsEvent.NoAssignmentViewed -> "no_assignment_viewed"
        is AnalyticsEvent.ApiError -> "api_error"
        is AnalyticsEvent.ApiRequestCompleted -> "api_request_completed"
        is AnalyticsEvent.PushNotificationReceived -> "push_notification_received"
        is AnalyticsEvent.PushNotificationTapped -> "push_notification_tapped"
        is AnalyticsEvent.SubtaskSelectSheetOpened -> "subtask_select_sheet_opened"
        is AnalyticsEvent.SubtaskSearchUsed -> "subtask_search_used"
        is AnalyticsEvent.SubtaskCreateSheetOpened -> "subtask_create_sheet_opened"
        is AnalyticsEvent.SubtaskCreated -> "subtask_created"
        is AnalyticsEvent.HintsTabViewed -> "hints_tab_viewed"
    }

    private fun AnalyticsEvent.params(): Map<String, String> = when (this) {
        is AnalyticsEvent.LoginCompleted -> mapOf("role" to role)
        is AnalyticsEvent.TasksListViewed -> mapOf("tasks_count" to tasksCount.toString())
        is AnalyticsEvent.TaskCardTapped -> mapOf("task_state" to taskState, "task_type" to taskType)
        is AnalyticsEvent.TaskDetailViewed -> mapOf(
            "task_state" to taskState,
            "task_review_state" to taskReviewState,
            "task_type" to taskType
        )
        is AnalyticsEvent.TaskCompleteSheetOpened -> mapOf("requires_photo" to requiresPhoto.toString())
        is AnalyticsEvent.TaskStartTapped -> buildMap {
            put("task_id", taskId)
            put("task_type", taskType)
            if (workType != null) put("work_type", workType)
            if (timeStart != null) put("time_start", timeStart)
            if (timeEnd != null) put("time_end", timeEnd)
        }
        is AnalyticsEvent.TaskCompleteSubmitted -> buildMap {
            put("has_photo", hasPhoto.toString())
            put("has_text", hasText.toString())
            put("task_id", taskId)
            put("task_type", taskType)
            if (workType != null) put("work_type", workType)
            if (timeStart != null) put("time_start", timeStart)
            if (timeEnd != null) put("time_end", timeEnd)
        }
        is AnalyticsEvent.ManagerHomeViewed -> mapOf("tasks_on_review_count" to tasksOnReviewCount.toString())
        is AnalyticsEvent.TaskReviewSheetOpened -> mapOf("task_review_state" to taskReviewState)
        is AnalyticsEvent.TaskRejectedTapped -> mapOf("has_comment" to hasComment.toString())
        is AnalyticsEvent.EmployeeFilterApplied -> mapOf("selected_count" to selectedCount.toString())
        is AnalyticsEvent.TaskCreateSubmitted -> mapOf(
            "has_assignee" to hasAssignee.toString(),
            "requires_photo" to requiresPhoto.toString(),
            "planned_minutes" to plannedMinutes.toString()
        )
        is AnalyticsEvent.ShiftOpenCompleted -> mapOf("shift_id" to shiftId.toString(), "role" to role)
        is AnalyticsEvent.ShiftCloseCompleted -> mapOf("shift_id" to shiftId.toString(), "role" to role)
        is AnalyticsEvent.ApiError -> mapOf("http_status" to httpStatus.toString())
        is AnalyticsEvent.ApiRequestCompleted -> buildMap {
            put("path", path)
            put("method", method)
            put("http_status", httpStatus.toString())
            put("duration_ms", durationMs.toString())
            put("store_id", storeId)
            put("is_error", isError.toString())
            if (errorType != null) put("error_type", errorType)
        }
        is AnalyticsEvent.PushNotificationReceived -> buildMap {
            put("channel", channel)
            if (notificationId != null) put("notification_id", notificationId)
            if (taskId != null) put("task_id", taskId)
        }
        is AnalyticsEvent.PushNotificationTapped -> buildMap {
            put("channel", channel)
            if (notificationId != null) put("notification_id", notificationId)
            if (taskId != null) put("task_id", taskId)
        }
        is AnalyticsEvent.SubtaskSelectSheetOpened -> buildMap {
            put("trigger", trigger)
            put("operations_count", operationsCount.toString())
            if (workType != null) put("work_type", workType)
        }
        is AnalyticsEvent.HintsTabViewed -> buildMap {
            put("hints_count", hintsCount.toString())
            if (workType != null) put("work_type", workType)
            if (zone != null) put("zone", zone)
        }
        else -> emptyMap()
    }
}
